framework/src/main/java/org/checkerframework/framework/util/typeinference8/constraint/ConstraintSet.java

```diff
@@ -153,35 +175,49 @@ public void remove(ConstraintSet subset) {
-   * each constraint, no input variable can influence an output variable of another constraint in
-   * this constraint set. (See JLS 18.5.2.2)
+   * A subset of constraints is selected in C, satisfying the property that, for each constraint, no
+   * input variable can influence an output variable of another constraint in C. (See JLS 18.5.2.2)
    *
+   * @param c a constraint set
    * @param dependencies an object describing the dependencies of inference variables
    * @return s a subset of constraints is this constraint set
    */
-  public ConstraintSet getClosedSubset(Dependencies dependencies) {
+  public static ConstraintSet getClosedSubset(ConstraintSet c, Dependencies dependencies) {
     ConstraintSet subset = new ConstraintSet();
-    Set<Variable> inputDependencies = new LinkedHashSet<>();
-    Set<Variable> outDependencies = new LinkedHashSet<>();
-    for (Constraint constraint : list) {
+    // Collection of all outputs of c.
+    Set<Variable> allOutputsOfC = new LinkedHashSet<>();
+    for (Constraint constraint : c.list) {
+      if (constraint instanceof TypeConstraint) {
+        allOutputsOfC.addAll(((TypeConstraint) constraint).getOutputVariables());
+      }
+      // No other constraints have output variables
+    }
+
+    // Find a subset of this set where the following is true for all the constraints in the subset:
+    // no input variable of a constraint can influence an output variable of another constraint in c
+    // (Influence means that neither variable can depend on the other.)
+    // The JLS does not specify whether this subset should be as large as possible, but this
+    // implementation returns the largest subset possible.
+    for (Constraint constraint : c.list) {
       if (constraint.getKind() == Kind.EXPRESSION
           || constraint.getKind() == Kind.LAMBDA_EXCEPTION
           || constraint.getKind() == Kind.METHOD_REF_EXCEPTION) {
-        TypeConstraint c = (TypeConstraint) constraint;
-        Set<Variable> newInputs = dependencies.get(c.getInputVariables());
-        Set<Variable> newOutputs = dependencies.get(c.getOutputVariables());
-        if (Collections.disjoint(newInputs, outDependencies)
-            && Collections.disjoint(newOutputs, inputDependencies)) {
-          inputDependencies.addAll(newInputs);
-          outDependencies.addAll(newOutputs);
-          subset.add(c);
-        } else {
-          // A cycle (or cycles) in the graph of dependencies between constraints exists.
-          subset = new ConstraintSet();
-          break;
+        List<Variable> inputsOfSingleConstraint = ((TypeConstraint) constraint).getInputVariables();
+        boolean foundInfluence = false;
+        for (Variable in : inputsOfSingleConstraint) {
+          for (Variable out : allOutputsOfC) {
+            if (dependencies.get(in).contains(out) || dependencies.get(out).contains(in)) {
+              foundInfluence = true;
+              break;
+            }
```

Comment:
I would add "if (foundInfluence) break;" here for efficiency, but maybe it doesn't matter.